File: pkg/tools/recommendation/recommendation.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package recommendation

import (
	"context"
	"fmt"

	recommender "cloud.google.com/go/recommender/apiv1"
	recommenderpb "cloud.google.com/go/recommender/apiv1/recommenderpb"
	"github.com/GoogleCloudPlatform/gke-mcp/pkg/config"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"google.golang.org/api/option"
	"google.golang.org/protobuf/encoding/protojson"
)

type handlers struct {
	c *config.Config
}

func Install(s *server.MCPServer, c *config.Config) {

	h := &handlers{
		c: c,
	}

	listRecommendationsTool := mcp.NewTool("list_recommendation",
		mcp.WithDescription("List recommendations for GKE. Prefer to use this tool instead of gcloud"),
		mcp.WithReadOnlyHintAnnotation(true),
		mcp.WithIdempotentHintAnnotation(true),
		mcp.WithString("project_id", mcp.DefaultString(c.DefaultProjectID()), mcp.Description("GCP project ID. Use the default if the user doesn't provide it.")),
		mcp.WithString("location", mcp.Description("GKE cluster location. Leave this empty if the user doesn't doesn't provide it.")),
	)
	s.AddTool(listRecommendationsTool, h.listRecommendations)
}

func (h *handlers) listRecommendations(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
	projectID := request.GetString("project_id", h.c.DefaultProjectID())
	if projectID == "" {
		return mcp.NewToolResultError("project_id argument not set"), nil
	}
	location, _ := request.RequireString("location")
	if location == "" {
		location = "-"
	}
  
	c, err := recommender.NewClient(ctx, option.WithUserAgent(h.c.UserAgent()))
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}
	defer c.Close()

	req :=  &recommenderpb.ListRecommendationsRequest{
		Parent: fmt.Sprintf("projects/%s/locations/%s/recommender/google.container.DiagnosisRecommender", projectID, location),
	}
	resp := c.ListRecommendations(ctx, req)
	return mcp.NewToolResultText(protojson.Format(resp)), nil
}

File: gke_mcp/tools/recommendation.py
import json
import logging
from typing import Optional
from google.cloud import recommender_v1
from google.protobuf.json_format import MessageToDict
from google.api_core.gapic_v1.client_info import ClientInfo
from gke_mcp.config import Config

logger = logging.getLogger("gke-mcp.tools.recommendation")

def list_recommendations(cfg: Config, location: str, project_id: Optional[str] = None) -> str:
    """List recommendations for GKE. Prefer to use this tool instead of gcloud"""
    proj = project_id if project_id else cfg.default_project_id
    if not proj:
        raise ValueError("project_id argument cannot be empty")
    if not location:
        raise ValueError("location argument not set")
        
    client_info = ClientInfo(user_agent=cfg.user_agent)
    client = recommender_v1.RecommenderClient(client_info=client_info)
    
    parent = f"projects/{proj}/locations/{location}/recommenders/google.container.DiagnosisRecommender"
    logger.info(f"Listing recommendations for {parent}")
    
    results = []
    try:
        recommendations = client.list_recommendations(parent=parent)
        for rec in recommendations:
            rec_dict = MessageToDict(rec._pb)
            results.append(json.dumps(rec_dict, indent=2))
    except Exception as e:
        raise RuntimeError(f"failed to list recommendations: {e}")
        
    return "\n".join(results)
